From a7ba8164d68c7029f99120f674388dea100b00ba Mon Sep 17 00:00:00 2001
From: Aris Wibowo <6703134+arswb@users.noreply.github.com>
Date: Sat, 10 May 2025 15:59:16 +0300
Subject: [PATCH] arm64/dts: sweet-sdmmagpie: Increase minimum backlight level
 again

[vbajs]
* Needed to fix 90hz lines at lowest brightness
* Value reversed from skiddie kernel dtbo

Signed-off-by: Yahya Wessam <yahyawessam2002@gmail.com>
---
 arch/arm64/boot/dts/qcom/xiaomi/sweet/sweet-sdmmagpie.dtsi | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/qcom/xiaomi/sweet/sweet-sdmmagpie.dtsi b/arch/arm64/boot/dts/qcom/xiaomi/sweet/sweet-sdmmagpie.dtsi
index 9f46b74ce474..993c22ccc040 100755
--- a/arch/arm64/boot/dts/qcom/xiaomi/sweet/sweet-sdmmagpie.dtsi
+++ b/arch/arm64/boot/dts/qcom/xiaomi/sweet/sweet-sdmmagpie.dtsi
@@ -721,7 +721,7 @@ And public attribution of xiaomi platforms(like F3 and so and)
 &dsi_k6_38_0c_0a_fhd_dsc_video {
 	qcom,panel-supply-entries = <&dsi_panel_pwr_supply_labibb_amoled>;
 	qcom,mdss-dsi-bl-pmic-control-type = "bl_ctrl_dcs";
-	qcom,mdss-dsi-bl-min-level = <8>;
+	qcom,mdss-dsi-bl-min-level = <20>;
 	qcom,mdss-dsi-bl-max-level = <2047>;
 	qcom,mdss-brightness-max-level = <2047>;
 	qcom,mdss-dsi-mode-sel-gpio-state = "single_port";

From 9507b1eb1c22076c193fc52d80907f0d588e69cc Mon Sep 17 00:00:00 2001
From: RooGhz720 <rooghz720@gmail.com>
Date: Sun, 15 Jan 2023 10:15:36 +0800
Subject: [PATCH] sweet: panel: add 90hz support

this initial beta and only work on miui 13 rom above with mod like magisk module to enable 90hz option

Signed-off-by: RooGhz720 <rooghz720@gmail.com>
Signed-off-by: Yahya Wessam <yahyawessam2002@gmail.com>
---
 .../qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi  | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi b/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi
index 344e06b7d4e1..aea728003ef0 100644
--- a/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi
+++ b/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi
@@ -58,7 +58,7 @@
 		/* DFPS */
 		qcom,mdss-dsi-pan-enable-dynamic-fps;
 		qcom,mdss-dsi-pan-fps-update = "dfps_immediate_porch_mode_vfp";
-		qcom,dsi-supported-dfps-list = <120 60>;
+		qcom,dsi-supported-dfps-list = <120 90 60>;
 
 		qcom,mdss-dsi-panel-dc-threshold = <488>;
 		qcom,mdss-dsi-panel-k6-dc-flag;

From e0e13496fbdc341e916cab2df48402672473bc5a Mon Sep 17 00:00:00 2001
From: RooGhz720 <rooghz720@gmail.com>
Date: Mon, 23 Jan 2023 12:40:08 +0800
Subject: [PATCH] arm64: adapt 90hz configurasi and recalculate lower
 brightness

Signed-off-by: RooGhz720 <rooghz720@gmail.com>
Signed-off-by: Yahya Wessam <yahyawessam2002@gmail.com>
---
 .../dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi    | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi b/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi
index d6e4c678c628..344e06b7d4e1 100644
--- a/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi
+++ b/arch/arm64/boot/dts/qcom/xiaomi/sweet/dsi-panel-k6-38-0c-0a-fhd-dsc-video.dtsi
@@ -177,6 +177,13 @@
 					39 00 00 00 00 00 02 F7 0B
 					39 01 00 00 00 00 03 F0 A5 A5];
 				qcom,mdss-dsi-dispparam-bc-120hz-command-state = "dsi_lp_mode";
+				qcom,mdss-dsi-dispparam-bc-90hz-command = [
+					/* Brightness Control - Freq Setting */
+					39 00 00 00 00 00 03 F0 5A 5A
+					39 00 00 00 00 00 02 60 01
+					39 00 00 00 00 00 02 F7 0B
+					39 01 00 00 00 00 03 F0 A5 A5];
+				qcom,mdss-dsi-dispparam-bc-90hz-command-state = "dsi_lp_mode";
 				qcom,mdss-dsi-dispparam-bc-60hz-command = [
 					/* Brightness Control - Freq Setting */
 					39 00 00 00 00 00 03 F0 5A 5A
@@ -204,6 +211,16 @@
 					39 01 00 00 00 00 03 51 01 E8
 				];
 				qcom,mdss-dsi-dispparam-60hz-dc-crc-setting-command-state = "dsi_lp_mode";
+				qcom,mdss-dsi-dispparam-90hz-dc-crc-setting-command = [
+					39 01 00 00 00 00 03 F0 5A 5A
+					39 00 00 00 00 00 02 80 05
+					39 00 00 00 00 00 02 B1 00
+					39 00 00 00 00 00 04 B0 00 01 B1
+					39 00 00 00 00 00 16 B1 01 00 00 00 01 00 00 00 01 00 01 01 01 00 01 01 01 00 01 01 01
+					39 00 00 00 00 00 03 F0 A5 A5
+					39 00 00 00 00 00 03 51 01 E8
+				];
+				qcom,mdss-dsi-dispparam-90hz-dc-crc-setting-command-state = "dsi_lp_mode";
 				qcom,mdss-dsi-dispparam-120hz-dc-crc-setting-command = [
 					39 01 00 00 00 00 03 F0 5A 5A
 					39 00 00 00 00 00 02 80 05

From 92fe8dd1f16e72d753faf224efc0ffa5b4d6a5a9 Mon Sep 17 00:00:00 2001
From: Yahya Wessam <yahyawessam2002@gmail.com>
Date: Sun, 12 Oct 2025 22:38:45 +0300
Subject: [PATCH] drm: msm: dsi-staging: hijack 60hz backlight control for 90hz
 on OSS drm

This fixes the high brightness and greenish tint when switching from 120hz to 90hz, but it doesn't fix the werid contrast

Signed-off-by: Yahya Wessam <yahyawessam2002@gmail.com>
---
 drivers/gpu/drm/msm/dsi-staging/dsi_display.c |  2 +-
 drivers/gpu/drm/msm/dsi-staging/dsi_panel.c   | 11 +++++++++--
 2 files changed, 10 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/msm/dsi-staging/dsi_display.c b/drivers/gpu/drm/msm/dsi-staging/dsi_display.c
index 49b3cb3a0820..daa94afb1c27 100644
--- a/drivers/gpu/drm/msm/dsi-staging/dsi_display.c
+++ b/drivers/gpu/drm/msm/dsi-staging/dsi_display.c
@@ -6962,7 +6962,7 @@ int dsi_display_set_mode(struct dsi_display *display,
 	}
 
 #ifdef CONFIG_MACH_XIAOMI_SWEET
-	if (adj_mode.timing.refresh_rate == 60)
+	if (adj_mode.timing.refresh_rate == 60 || adj_mode.timing.refresh_rate == 90)
 		dsi_panel_gamma_mode_change(display->panel, &adj_mode);
 #endif
 
diff --git a/drivers/gpu/drm/msm/dsi-staging/dsi_panel.c b/drivers/gpu/drm/msm/dsi-staging/dsi_panel.c
index fe315344256e..7652d9e09a55 100644
--- a/drivers/gpu/drm/msm/dsi-staging/dsi_panel.c
+++ b/drivers/gpu/drm/msm/dsi-staging/dsi_panel.c
@@ -3864,10 +3864,17 @@ void dsi_panel_gamma_mode_change(struct dsi_panel *panel,
 	if (!count)
 		goto exit;
 
-	if (adj_mode->timing.refresh_rate == 120)
+	switch (adj_mode->timing.refresh_rate) {
+	case 120:
 		rc = dsi_panel_tx_cmd_set(panel, DSI_CMD_SET_DISP_BC_120HZ);
-	else if (adj_mode->timing.refresh_rate == 60)
+		break;
+	case 90:
+	case 60:
 		rc = dsi_panel_tx_cmd_set(panel, DSI_CMD_SET_DISP_BC_60HZ);
+		break;
+	default:
+		break;
+	}
 
 	if (rc)
 		pr_err("%s: send cmds failed...", __func__);

